py-uxarray: add 2025.5.x and fix dependencies

Chrismarsh opened the pull request:
- This adds `py-uxarray@2025.5.0`
- Updates a bunch of the dependencies 

Requires:
-  https://github.com/spack/spack/pull/50075. Required for all the `py-numpy@2` fixes.
-  https://github.com/spack/spack/pull/50415 rust changes to fix version string

Chrismarsh committed:
add new uxarray version and fix dependencies

climbfuji commented:
@cmarsh This doesn't concretize for me:
```
$ git clone -b package-fix/py-uxarray https://github.com/Chrismarsh/spack spack-cmarsh-py-uxarray
Cloning into 'spack-cmarsh-py-uxarray'...
remote: Enumerating objects: 629868, done.
remote: Counting objects: 100% (318/318), done.
remote: Compressing objects: 100% (193/193), done.
remote: Total 629868 (delta 239), reused 137 (delta 116), pack-reused 629550 (from 3)
Receiving objects: 100% (629868/629868), 215.12 MiB | 3.87 MiB/s, done.
Resolving deltas: 100% (297930/297930), done.
Updating files: 100% (12267/12267), done.

$ cd spack-cmarsh-py-uxarray

$ source share/spack/setup-env.sh

$ spack env create py-uxarray-gcc-test
==> Created environment py-uxarray-gcc-test in: /home/dom/scratch/spack-cmarsh-py-uxarray/var/spack/environments/py-uxarray-gcc-test
==> Activate with: spack env activate py-uxarray-gcc-test

$ spack env activate -p py-uxarray-gcc-test

$ spack config edit
# added "py-uxarray" to empty list of specs

$ spack external find
==> The following specs have been detected on this system and added to /home/dom/scratch/spack-cmarsh-py-uxarray/var/spack/environments/py-uxarray-gcc-test/spack.yaml
autoconf@2.69    binutils@2.35.2  cmake@3.26.5    curl@7.76.1    findutils@4.8.0  gettext@0.21  git-lfs@3.4.1  groff@1.22.4  openssh@8.7p1  perl@5.32.1    python@3.9.21  tar@1.34
automake@1.16.2  bison@3.7.4      coreutils@8.32  diffutils@3.7  gawk@5.1.0       git@2.43.5    gmake@4.3      m4@1.4.19     openssl@3.2.2  pkgconf@1.7.3  sed@4.8        zlib@1.2.11

$ spack compiler find
==> Added 1 new compiler to /home/dom/scratch/spack-cmarsh-py-uxarray/var/spack/environments/py-uxarray-gcc-test/spack.yaml
    gcc@11.5.0
==> Compilers are defined in the following files:
    /home/dom/scratch/spack-cmarsh-py-uxarray/var/spack/environments/py-uxarray-gcc-test/spack.yaml

$ spack concretize --force --fresh 2>&1 | tee log.concretize.py-uxarray-gcc-test.001
==> Error: failed to concretize `py-uxarray` for the following reasons:
     1. Cannot satisfy 'py-dask@2024.7.1:'
        required because py-uxarray depends on py-dask@2024.7.1:+dataframe
          required because py-uxarray requested explicitly
     2. Cannot satisfy 'py-dask@2024.7.1:'

```

Chrismarsh committed:
style

climbfuji commented:
Thanks. AFter pulling in 50075, it concretizes, and it is now running the install. We'll know tomorrow if it succeeded.

Chrismarsh committed:
fix py bokeh range and add new py-bokeh

Chrismarsh committed:
new version for py bokeh added last commit

Chrismarsh commented:
@climbfuji Great thanks! I just added a new `py-bokeh` version and fixed a version error of it in `py-panel` that the CI caught. I don't think it would impact your build but just FYI

Chrismarsh committed:
add missing setuptools

Chrismarsh committed:
add conflict on gcc11 for 1.20:1.25 as per https://github.com/spack/spack/pull/50351

Chrismarsh committed:
update gcc conflict range

Chrismarsh commented:
@climbfuji 
I've made the additions. It looks like 1.20, 1.21, 1.22 also has known issues with gcc@11 so I've refined the conflict 

https://github.com/numpy/numpy/pull/21536
https://github.com/numpy/numpy/releases/tag/v1.21.1

I have also added the missing setuptools.

climbfuji commented on var/spack/repos/spack_repo/builtin/packages/py_pyogrio/package.py:
```diff
@@ -16,22 +16,21 @@ class PyPyogrio(PythonPackage):
-    depends_on("gdal@2.4:", type=("build", "link", "run"))
-    depends_on("py-cython@0.29:", type="build")
 
+    depends_on("py-setuptools", type=("build"))
```
```
depends_on("py-setuptools", type="build")
```

climbfuji commented on var/spack/repos/spack_repo/builtin/packages/py_numpy/package.py:
```diff
@@ -242,9 +242,12 @@ class PyNumpy(PythonPackage):
 
     # gcc 12 has the correct avx512 implimentation
```
implementation

climbfuji commented:
Thanks. I was able to build it, but I can't import the uxarray module because of the following error. Is this because I also need to other PR #50352? It looks different, though (`dask-expr` missing).
```
[py-uxarray-gcc-test] [dom@bounty spack-cmarsh-py-uxarray]$ python3
Python 3.11.11 (main, May  7 2025, 20:31:17) [GCC 11.5.0 20240719 (Red Hat 11.5.0-5)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import uxarray
/home/dom/scratch/spack-cmarsh-py-uxarray/var/spack/environments/py-uxarray-gcc-test/.spack-env/view/lib/python3.11/site-packages/dask/dataframe/__init__.py:42: FutureWarning:
Dask dataframe query planning is disabled because dask-expr is not installed.

You can install it with `pip install dask[dataframe]` or `conda install dask`.
This will raise in a future version.

  warnings.warn(msg, FutureWarning)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/dom/scratch/spack-cmarsh-py-uxarray/var/spack/environments/py-uxarray-gcc-test/.spack-env/view/lib/python3.11/site-packages/uxarray/__init__.py", line 3, in <module>
    from .core.api import concat, open_dataset, open_grid, open_mfdataset
  File "/home/dom/scratch/spack-cmarsh-py-uxarray/var/spack/environments/py-uxarray-gcc-test/.spack-env/view/lib/python3.11/site-packages/uxarray/core/__init__.py", line 1, in <module>
    from .api import open_dataset, open_grid, open_mfdataset
  File "/home/dom/scratch/spack-cmarsh-py-uxarray/var/spack/environments/py-uxarray-gcc-test/.spack-env/view/lib/python3.11/site-packages/uxarray/core/api.py", line 10, in <module>
    from uxarray.core.dataset import UxDataset
  File "/home/dom/scratch/spack-cmarsh-py-uxarray/var/spack/environments/py-uxarray-gcc-test/.spack-env/view/lib/python3.11/site-packages/uxarray/core/dataset.py", line 16, in <module>
    from uxarray.core.dataarray import UxDataArray
  File "/home/dom/scratch/spack-cmarsh-py-uxarray/var/spack/environments/py-uxarray-gcc-test/.spack-env/view/lib/python3.11/site-packages/uxarray/core/dataarray.py", line 16, in <module>
    from uxarray.core.aggregation import _uxda_grid_aggregate
  File "/home/dom/scratch/spack-cmarsh-py-uxarray/var/spack/environments/py-uxarray-gcc-test/.spack-env/view/lib/python3.11/site-packages/uxarray/core/aggregation.py", line 5, in <module>
    from uxarray.grid.connectivity import get_face_node_partitions
  File "/home/dom/scratch/spack-cmarsh-py-uxarray/var/spack/environments/py-uxarray-gcc-test/.spack-env/view/lib/python3.11/site-packages/uxarray/grid/__init__.py", line 1, in <module>
    from .grid import Grid
  File "/home/dom/scratch/spack-cmarsh-py-uxarray/var/spack/environments/py-uxarray-gcc-test/.spack-env/view/lib/python3.11/site-packages/uxarray/grid/grid.py", line 71, in <module>
    from uxarray.grid.validation import (
  File "/home/dom/scratch/spack-cmarsh-py-uxarray/var/spack/environments/py-uxarray-gcc-test/.spack-env/view/lib/python3.11/site-packages/uxarray/grid/validation.py", line 4, in <module>
    import polars as pl
ModuleNotFoundError: No module named 'polars'
>>>
```

Chrismarsh commented:
Because the list of depends was in a different order than the `pyproject.toml` I missed that `py-polars` was a required dependency. This is simple addition is turning into a huge pain because of polars seemingly requiring features that only exist in nightly rust.

Chrismarsh committed:
missing depends, reorder list to match pyproject.toml

Chrismarsh committed:
add new package healpix for uxarray dependency

Chrismarsh committed:
update dependencies

Chrismarsh committed:
update polars

Chrismarsh commented:
This now has all the missing dependencies and an updated `py-polars`. I don't love that it had to depend on `rust@nightly` which is a bit of a pain to install. This may need further patches to fix `argminmax` if 0.6.3 doesn't fix it -- testing now.

climbfuji commented:
Depending on `rust@nightly` is an absolute showstopper. We will not use `uxarray` until this is fixed.

Chrismarsh commented:
Did `@0.20` work with non-nightly rust? `2025` certainly doesn't -- it does not compile with a non-nightly rust. `polars` seems to be pretty clear in that it doesn't work with non-nightly rust
https://github.com/pola-rs/polars/issues/13653#issuecomment-2041619141

>We will not use uxarray until this is fixed.

We = ?

climbfuji commented:
we = spack-stack developers and users (most of NOAA EPIC/UFS, Navy, JCSDA, ...)

Chrismarsh commented:
Caveat: I don't do rust work and have no idea about rust.

I noticed the non-nightly rust compiler identifies itself as a nightly
```
rustc --version
rustc 1.85.0-nightly (4d91de4e4 2025-02-17) (built from a source tarball)
```
which seems to trigger the nightly code paths. Maybe fixing that will let this work with a stable rust

climbfuji commented:
I wish rust and cargo had never been introduced in the Python world. Sigh.

Chrismarsh committed:
style

Chrismarsh committed:
style

Chrismarsh committed:
switch to build run to follow the python default

Chrismarsh committed:
update to patch out nightly reqs of dependent crates

Chrismarsh commented:
Indeed there is a problem with the stable rusts -- they have a `-nightly` suffix in the version which triggers code paths reserved for nightly compilers. This PR now depends on the rust changes in https://github.com/spack/spack/pull/50415 which fixes this.

With the rust fix and the update to this PR, `py-uxarray` now builds with a stable rust. It does so by removing the nightly features of the dependent crates. `py-polars` proper seems to protect all its nightly code paths.

Chrismarsh commented:
I am confirming this approach here 
https://github.com/pola-rs/polars/issues/22708

Chrismarsh committed:
style

Chrismarsh committed:
style in regex

Chrismarsh committed:
style

Chrismarsh commented:
@climbfuji Urgh, I have no idea, this PR doesn't touch `arrow` or `thrift` and it "builds for me" (tm).

I suspect it's that my `^thrift` is `build_system=cmake` whereas yours is `build_system=autotools`. I've seen this problem in other libraries. The cmake built-libraries expect cmake config files, autoools expects pkg-config. If you build `thrift` with cmake does it work (I have a setting to prefer cmake globally to avoid this problem)?

Indeed I tried to build `arrow ^thrift build_system=autotools` but it fails to configure with `configure: error: Invalid --with-openssl value`. The package.py looks broken on that line but I'm not sure what it should be
https://github.com/spack/spack/blob/develop/var/spack/repos/spack_repo/builtin/packages/thrift/package.py#L136. In your clearly successful build, what is the value?

Chrismarsh commented:
I ended up removing that line of `package.py` to force it to configure, and `arrow buildsystem=cmake ^thrift build_system=autotools` worked for me.  So that's not it. I am using `thrift@0.21.0` -- can you try that instead of `@0.20`? I can't find much of a change log between them. I run into this from time to time, https://github.com/spack/spack/issues/45983, where I have to deactivate then re activate the env for it to pick up paths correctly.

climbfuji commented:
> I ended up removing that line of `package.py` to force it to configure, and `arrow buildsystem=cmake ^thrift build_system=autotools` worked for me. So that's not it. I am using `thrift@0.21.0` -- can you try that instead of `@0.20`? I can't find much of a change log between them. I run into this from time to time, #45983, where I have to deactivate then re activate the env for it to pick up paths correctly.

I have thrift 0.21.0. Will check later if I can switch the buildsystem.

climbfuji commented:
@Chrismarsh I forced spack to use `thrift@0.21.0 build_system=cmake`. All the dependencies built correctly, and I can run this:
```
[py-uxarray-gcc-test] [dom@bounty spack-cmarsh-py-uxarray]$ python3
Python 3.11.11 (main, May  7 2025, 20:31:17) [GCC 11.5.0 20240719 (Red Hat 11.5.0-5)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import uxarray
>>>
```

Chrismarsh committed:
add new version

Chrismarsh committed:
add nightly variant. pass arguments to maturin instead of patching

Chrismarsh committed:
ensure arrow +parquet uses ^thrift cmake so that it is found

Chrismarsh commented:
@climbfuji amazing thanks! I have received feedback on the maturin args needed for `py-polars`, so I have switched from the patch to passing the arguments directly. I have also enforced `^thrift build_system=cmake` for `arrow` given it impacts this PR.

Chrismarsh committed:
typo

Chrismarsh committed:
style

climbfuji commented on var/spack/repos/spack_repo/builtin/packages/py_holoviews/package.py:
```diff
@@ -14,8 +14,8 @@ class PyHoloviews(PythonPackage):
 
```
You can put me down as maintainer for holoviews if you like

Chrismarsh replied:
added

climbfuji commented on var/spack/repos/spack_repo/builtin/packages/py_pyogrio/package.py:
```diff
@@ -16,15 +16,21 @@ class PyPyogrio(PythonPackage):
 
```
same here, can add me as maintainer if you like

Chrismarsh replied:
you already are a maintainer on this one

climbfuji reviewed (approved):
This works for me, thanks very much for all your work @Chrismarsh

Chrismarsh committed:
add maintainer

Chrismarsh committed:
note on missing dependencies

Chrismarsh commented:
Did you need the rust changes in https://github.com/spack/spack/pull/50415 to compile this?

Also: I have only added some of the optional dependencies listed here
https://github.com/pola-rs/polars/blob/py-1.29.0/py-polars/pyproject.toml#L42

The db and excel dependencies that I didn't include are all missing in spack. I am not adding them now to keep the size of the PR down and I also suspect they are not the heavily used parts of polars in spack.

Chrismarsh commented:
@climbfuji given you needed it to test this PR, are you open to reviewing https://github.com/spack/spack/pull/50075 too?

Chrismarsh commented:
I didn't change anything with Thrift, so I don't understand how that can be the case. Especially since you built Thrift with cmake just yesterday! I do see that Thrift doesn't actually depend on boost
https://github.com/spack/spack/blob/develop/var/spack/repos/spack_repo/builtin/packages/thrift/package.py
despite it totally depending on boost
https://github.com/apache/thrift/blob/master/CMakeLists.txt#L63

does adding a depends on boost fix it?

Chrismarsh committed:
add missing boost depends

Chrismarsh committed:
code path seems to require boost even when ~cpp

Chrismarsh committed:
fix missing import

Chrismarsh committed:
missing import
